File: distancedetectorjava/src/main/java/com/example/distancedetectorjava/DistanceDetector.java
package com.example.distancedetectorjava;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.widget.Toast;

import androidx.core.content.ContextCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import java.util.ArrayList;

public class DistanceDetector {

    public static double lat,lng,userDistance;
    public static String status = "";

    boolean checkPreviousLatLng = false;
    public double distance;
    static LocationModel locationModel = new LocationModel();

    static Context context;
    private DistanceListener distanceListener;
    public void setDistanceListener(DistanceListener mSpeedChangeListener) {
        this.distanceListener = mSpeedChangeListener;
    }



    public DistanceDetector(Context context) {
        this.context = context;
    }

    public static double getLatitude(){
        return lat;
    }
    public static double getLongitude(){
        return lng;
    }
    public static double getUserDistance(){
        return userDistance;
    }
    public static void setUserDistance(double uDistance){
        userDistance = uDistance;
    }

    public interface DistanceListener {
        //public void onDistanceChange(double lat, double lng, double distance, String status);
        public void onDistanceChange(String status);
    }

    public double distance(double lat1, double lng1, double lat2, double lng2) {
        double earthRadius = 6371000; //meters
        double dLat = Math.toRadians(lat2-lat1);
        double dLng = Math.toRadians(lng2-lng1);
        double a = Math.sin(dLat/2) * Math.sin(dLat/2) +
                Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                        Math.sin(dLng/2) * Math.sin(dLng/2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));

        return (earthRadius * c);
    }


    private BroadcastReceiver locationReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(final Context context, Intent intent) {

            if(!checkPreviousLatLng){
                locationModel.setpLat(intent.getDoubleExtra(LocationConstants.INTENT_KEY_LATITUDE, 0));
                locationModel.setpLng(intent.getDoubleExtra(LocationConstants.INTENT_KEY_LONGITUDE, 0));

                checkPreviousLatLng = true;

            }else{
                locationModel.setcLat(intent.getDoubleExtra(LocationConstants.INTENT_KEY_LATITUDE, 0));
                locationModel.setcLng(intent.getDoubleExtra(LocationConstants.INTENT_KEY_LONGITUDE, 0));


                distance = distance(locationModel.getpLat(),locationModel.getpLng(),locationModel.getcLat(),locationModel.getcLng());

            }
            lat = intent.getDoubleExtra(LocationConstants.INTENT_KEY_LATITUDE,0);
            lng = intent.getDoubleExtra(LocationConstants.INTENT_KEY_LONGITUDE,0);


            if (distance > userDistance){
                distanceListener.onDistanceChange("out of distance");
            }else{
                //distanceListener.onDistanceChange("safe ");
            }



            }
    };

    public void registerLocationReceiver() {
        LocalBroadcastManager.getInstance(context).registerReceiver(
                locationReceiver, new IntentFilter(LocationConstants.LIVE_LOCATION_BROADCAST_CHANNEL));
    }

    public void unregisterLocationReceiver() {
        try{
            LocalBroadcastManager.getInstance(context).unregisterReceiver(locationReceiver);
        }catch (Exception e){
            Toast.makeText(context, e.getLocalizedMessage(), Toast.LENGTH_LONG).show();
        }

    }

    public static void startService(Context context) {
        Intent currentLocationService = new Intent(context, CurrentLocationSendingService.class);
        Intent distanceStatus = new Intent(context, DistanceDetectorService.class);
        ContextCompat.startForegroundService(context, distanceStatus);
        ContextCompat.startForegroundService(context, currentLocationService);

    }

    public static void stopService(){
        context.stopService(new Intent(context, CurrentLocationSendingService.class));
        context.stopService(new Intent(context, DistanceDetectorService.class));
    }


}

File: distancedetector/src/main/java/com/example/distancedectector/DistanceDetector.kt
package com.example.distancedectector

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import java.util.ArrayList

class DistanceDetector(applicationContext: Context) {
    companion object {
        private var lat: Double = 0.0
        private var lng: Double = 0.0
        private var userDistance: Double = 0.0
        //private var context: Context? = null


        fun getLatitude(): Double {
            return lat
        }

        fun getLongitude(): Double {
            return lng
        }

        fun getUserDistance(): Double {
            return userDistance
        }

        fun setUserDistance(uDistance: Double) {
            userDistance = uDistance
        }

        fun startService(context: Context) {
            val currentLocationService = Intent(context, CurrentLocationSendingService::class.java)
            val distanceStatus = Intent(context, DistanceDetectorService::class.java)
            ContextCompat.startForegroundService(context, distanceStatus)
            ContextCompat.startForegroundService(context, currentLocationService)
            registerStatusReceiver(context)

        }

        fun stopService(context: Context) {
            context.stopService(Intent(context, CurrentLocationSendingService::class.java))
            context.stopService(Intent(context, DistanceDetectorService::class.java))
            unregisterStatusReceiver(context)

        }

        private var statusReceiver: BroadcastReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                //status = intent.getStringExtra(LocationConstants.INTENT_KEY_STATUS)
            }
        }

        private fun registerStatusReceiver(context: Context) {
            LocalBroadcastManager.getInstance(context).registerReceiver(
                statusReceiver, IntentFilter(LocationConstants.LIVE_LOCATION_BROADCAST_CHANNEL)
            )

        }

        private fun unregisterStatusReceiver(context: Context) {

            LocalBroadcastManager.getInstance(context).unregisterReceiver(statusReceiver)

        }
    }

    private var checkPreviousLatLng = false
    var distance: Double = 0.toDouble()
    private var locationModel = LocationModel()

    private lateinit var context: Context
    private lateinit var distanceListener: DistanceListener


    fun setDistanceListener(mSpeedChangeListener: DistanceListener) {
        this.distanceListener = mSpeedChangeListener
    }


    interface DistanceListener {
        //public void onDistanceChange(double lat, double lng, double distance, String status);
        fun onDistanceChange(status: String)

    }


    fun distance(lat1: Double, lng1: Double, lat2: Double, lng2: Double): Double {
        val earthRadius = 6371000.0 //meters
        val dLat = Math.toRadians(lat2 - lat1)
        val dLng = Math.toRadians(lng2 - lng1)
        val a =
            Math.sin(dLat / 2) * Math.sin(dLat / 2) + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                    Math.sin(dLng / 2) * Math.sin(dLng / 2)
        val c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))

        return earthRadius * c
    }


    private val locationReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {

            if (!checkPreviousLatLng) {
                locationModel.pLat = (intent.getDoubleExtra(LocationConstants.INTENT_KEY_LATITUDE, 0.0))
                locationModel.pLng = (intent.getDoubleExtra(LocationConstants.INTENT_KEY_LONGITUDE, 0.0))

                checkPreviousLatLng = true

            } else {
                locationModel.cLat = (intent.getDoubleExtra(LocationConstants.INTENT_KEY_LATITUDE, 0.0))
                locationModel.cLng = (intent.getDoubleExtra(LocationConstants.INTENT_KEY_LONGITUDE, 0.0))



                distance = distance(
                    locationModel.pLat,
                    locationModel.pLng,
                    locationModel.cLat,
                    locationModel.cLng
                )
            }
            lat = intent.getDoubleExtra(LocationConstants.INTENT_KEY_LATITUDE, 0.0)
            lng = intent.getDoubleExtra(LocationConstants.INTENT_KEY_LONGITUDE, 0.0)
            //            lat = locationModel.getcLat();
            //            lng = locationModel.getcLng();

            if (distance > userDistance) {
                distanceListener.onDistanceChange("out of limit $distance")
            }/* else {
                distanceListener.onDistanceChange("safe $distance")
            }*/


        }
    }

    fun registerLocationReceiver(context: Context) {
        LocalBroadcastManager.getInstance(context).registerReceiver(
            locationReceiver, IntentFilter(LocationConstants.LIVE_LOCATION_BROADCAST_CHANNEL)
        )

    }

    fun unregisterLocationReceiver() {
        try {
            LocalBroadcastManager.getInstance(context).unregisterReceiver(locationReceiver)
        } catch (e: Exception) {
            Toast.makeText(context, e.localizedMessage, Toast.LENGTH_LONG).show()
        }

    }


}